File: forensics/hops.py
"""
Multi-hop money tracing.

BFS traversal: starting from an address, follow outflows N hops deep.
Each hop investigates a new address via Etherscan and stores flows to DuckDB.
"""

import asyncio
from collections import defaultdict
from indexer.etherscan import investigate_address
from db.database import store_flows


async def trace_hops(
    start_address: str,
    chain: str,
    max_hops: int = 3,
    min_amount: float = 0.1,
) -> dict:
    """
    Follow money from start_address up to max_hops deep.

    Example tree:
        Hop 1: start → A, start → B
        Hop 2: A → C, A → D, B → E
        Hop 3: C → F, D → G

    Args:
        start_address: Ethereum address to start from.
        chain:         Chain name (must be in config/chains.py).
        max_hops:      Maximum depth to traverse.
        min_amount:    Skip edges with amount below this threshold.

    Returns:
        dict with graph edges and all addresses found.
    """
    visited: set[str] = set()
    graph:   list[dict] = []
    queue:   list[tuple[str, int]] = [(start_address.lower(), 0)]

    while queue:
        address, hop = queue.pop(0)
        if address in visited or hop >= max_hops:
            continue
        visited.add(address)

        print(f"  Hop {hop + 1}: investigating {address}")
        try:
            data = await investigate_address(address, chain)
        except Exception as e:
            print(f"  [hops] error at {address}: {e}")
            continue

        stored = store_flows(data["all_flows"])
        print(f"  └─ stored {stored} flows")

        for flow in data["outflows"]:
            if flow["amount"] < min_amount:
                continue
            destination = flow["to_address"]
            graph.append({
                "source":      address,
                "destination": destination,
                "chain":       chain,
                "hop_number":  hop + 1,
                "token":       flow["token"],
                "amount":      flow["amount"],
                "tx_hash":     flow["tx_hash"],
            })
            if destination not in visited:
                queue.append((destination, hop + 1))

    return {
        "start":           start_address,
        "chain":           chain,
        "hops":            max_hops,
        "graph":           graph,
        "addresses_found": list(visited),
    }


def summarise_hop_graph(graph: list[dict]) -> dict:
    """
    Aggregate the hop graph by destination address.

    Returns:
        Dict mapping address → {total_received, tx_count, tokens}
        sorted by total_received descending.
    """
    volume = defaultdict(float)
    count  = defaultdict(int)
    tokens: dict[str, set] = defaultdict(set)

    for edge in graph:
        dest = edge["destination"]
        volume[dest] += edge["amount"]
        count[dest]  += 1
        tokens[dest].add(edge["token"])

    return {
        addr: {
            "total_received": round(volume[addr], 6),
            "tx_count":       count[addr],
            "tokens":         list(tokens[addr]),
        }
        for addr in sorted(volume, key=lambda x: -volume[x])
    }


def build_networkx_graph(graph: list[dict]):
    """
    Convert the hop graph list into a NetworkX DiGraph for visualisation.
    Requires: pip install networkx
    """
    try:
        import networkx as nx
        G = nx.DiGraph()
        for edge in graph:
            G.add_edge(
                edge["source"],
                edge["destination"],
                weight=edge["amount"],
                token=edge["token"],
                hop=edge["hop_number"],
            )
        return G
    except ImportError:
        raise RuntimeError("networkx not installed — run: pip install networkx")

File: src/lib/forensics/hops.ts
import { investigateAddress, fetchNativeTransfers, fetchInternalTransfers } from "../etherscan";
import { storeFlows, queryRows } from "../db";

export interface HopEdge {
  source: string;
  destination: string;
  chain: string;
  hop_number: number;
  token: string;
  amount: number;
  tx_hash: string;
}

export interface TraceHopsResult {
  start: string;
  chain: string;
  hops: number;
  graph: HopEdge[];
  addresses_found: string[];
}

export async function traceHops(
  startAddress: string,
  chain: string,
  maxHops = 3,
  minAmount = 0.1
): Promise<TraceHopsResult> {
  const visited = new Set<string>();
  const graph: HopEdge[] = [];
  const queue: { address: string; hop: number }[] = [
    { address: startAddress.toLowerCase(), hop: 0 },
  ];

  while (queue.length > 0) {
    const { address, hop } = queue.shift()!;
    if (visited.has(address) || hop >= maxHops) {
      continue;
    }
    visited.add(address);

    console.log(`  Hop ${hop + 1}: investigating ${address}`);
    try {
      const data = await investigateAddress(address, chain);
      const stored = await storeFlows(data.all_flows);
      console.log(`  └─ stored ${stored} flows`);

      for (const flow of data.outflows) {
        if (flow.amount < minAmount) continue;

        const destination = flow.to_address.toLowerCase();
        graph.push({
          source: address,
          destination,
          chain,
          hop_number: hop + 1,
          token: flow.token,
          amount: flow.amount,
          tx_hash: flow.tx_hash,
        });

        if (!visited.has(destination)) {
          queue.push({ address: destination, hop: hop + 1 });
        }
      }
    } catch (e) {
      console.error(`  [hops] error at ${address}:`, e);
    }
  }

  return {
    start: startAddress,
    chain,
    hops: maxHops,
    graph,
    addresses_found: Array.from(visited),
  };
}

export interface SummaryInfo {
  total_received: number;
  tx_count: number;
  tokens: string[];
}

export function summariseHopGraph(graph: HopEdge[]): Record<string, SummaryInfo> {
  const volume: Record<string, number> = {};
  const count: Record<string, number> = {};
  const tokens: Record<string, Set<string>> = {};

  for (const edge of graph) {
    const dest = edge.destination;
    volume[dest] = (volume[dest] || 0) + edge.amount;
    count[dest] = (count[dest] || 0) + 1;
    if (!tokens[dest]) {
      tokens[dest] = new Set<string>();
    }
    tokens[dest].add(edge.token);
  }

  const sortedAddresses = Object.keys(volume).sort((a, b) => volume[b] - volume[a]);

  const summary: Record<string, SummaryInfo> = {};
  for (const addr of sortedAddresses) {
    summary[addr] = {
      total_received: Math.round(volume[addr] * 1000000) / 1000000,
      tx_count: count[addr],
      tokens: Array.from(tokens[addr]),
    };
  }

  return summary;
}

export interface FunderResult {
  wallet: string;
  chain: string;
  funder: string;
  amount: number;
  token: string;
  tx_hash: string;
  block_time: string;
}

export async function getFirstFunder(wallet: string, chain: string): Promise<FunderResult | null> {
  const cleanWallet = wallet.toLowerCase();

  // 1. Try to find in database first
  const rows = await queryRows(
    `SELECT from_address, token, amount, tx_hash, block_time FROM address_flows
     WHERE to_address = ?
       AND chain = ?
       AND (token LIKE '%ETH%' OR token LIKE '%MATIC%' OR token LIKE '%BNB%' OR token LIKE '%AVAX%')
     ORDER BY block_time ASC
     LIMIT 1`,
    [cleanWallet, chain]
  );

  if (rows.length > 0) {
    return {
      wallet: cleanWallet,
      chain,
      funder: rows[0].from_address,
      amount: rows[0].amount,
      token: rows[0].token,
      tx_hash: rows[0].tx_hash,
      block_time: rows[0].block_time,
    };
  }

  // 2. If not found in database, fetch from explorer
  console.log(`[funder] fetching funding history from explorer for ${cleanWallet} on ${chain}`);
  const nativeTransfers = await fetchNativeTransfers(cleanWallet, chain);
  const internalTransfers = await fetchInternalTransfers(cleanWallet, chain);

  const allTransfers = [...nativeTransfers, ...internalTransfers];
  if (allTransfers.length === 0) {
    return null;
  }

  // Save the fetched transfers to database so we cache them
  await storeFlows(allTransfers);

  // Find the first incoming transfer where to_address is cleanWallet
  const incoming = allTransfers
    .filter((t) => t.to_address === cleanWallet)
    .sort((a, b) => new Date(a.block_time).getTime() - new Date(b.block_time).getTime());

  if (incoming.length === 0) {
    return null;
  }

  const first = incoming[0];
  return {
    wallet: cleanWallet,
    chain,
    funder: first.from_address,
    amount: first.amount,
    token: first.token,
    tx_hash: first.tx_hash,
    block_time: first.block_time,
  };
}
